docs/_posts/2022-05-06-writing-cli-tools-with-typescript.md

```diff
@@ -0,0 +1,194 @@
+* [Click](https://docs.python-guide.org/scenarios/cli/) (Python)
+* [Picocli](https://picocli.info/picocli-2.0-groovy-scripts-on-steroids.html) (Groovy)
+* [Commander.js](https://github.com/tj/commander.js) (Node.js)
+
+
+
+So, why would we prefer writing a CLI in TypeScript given all these other choices?
+
+Well, TypeScript is an overall good choice for a couple of reasons: 
+
+1. **compatibility** with JavaScript (TypeScript is a superset of—and by that token compatible with—JavaScript),
+2. **popularity** (TypeScript was ranked 7th in the [2021 Stack Overflow Developer Survey](https://insights.stackoverflow.com/survey/2021#section-most-popular-technologies-programming-scripting-and-markup-languages) with JavaScript being 1st),
+3. **tooling choice** (the Node.js ecosystem provides [quite a lot of CLI libraries](https://openbase.com/categories/js/best-javascript-cli-libraries)),
+4. **developer friendliness** (TypeScript is a scripted language but also has an advanced type system and static type checking which helps writing high quality code),
+5. **performance** (since TypeScript code is in the end typically run by Node.js it is in general sufficiently performant thanks to the [v8](https://v8.dev/) engine).
+
+
+
+Nonetheless, a typically non-trivial part in any TypeScript project is set up. Using e.g. Python with the [Click](https://click.palletsprojects.com/en/8.1.x/) library, requires little more than writing a file with a [main function](https://docs.python.org/3/library/__main__.html#packaging-considerations). This set up complexity is of course due to the nature of TypeScript as a superset of JavaScript. To run TypeScript code, you typically transpile to JavaScript and then execute the JavaScript code with Node.js. (Yes, even [`ts-node`](https://github.com/TypeStrong/ts-node) does that.) So, before running a program written in TypeScript, we we need to transpile to JavaScript first. This requires tooling and therefore setup.
+
+
+
+Preferably, we also want to have a single, packaged binary as the result of building our project rather than a script file to reduce installation overhead on part of the user. 
+
+To assure high code quality we also typically want to include code formatting, linting, and tests.
+
+## Basic development flow
+
+To hit the ground running, we use a starter project which already contains all the set up required for running our code, formatting, linting, testing and packaging to a binary and install its dependencies. 
+
+> :information_source: The starter project for can be found on my [GitHub](https://github.com/fkurz/typescript-cliutil-starter) and on [npm](https://www.npmjs.com/package/@fkurz/typescript-cliutil-starter/). 
+
+```bash
+git clone git@github.com:fkurz/typescript-cliutil-starter.git my-cli
+cd my-cli 
+npm i
+```
+
+Our starter project typescript-cliutil-starter uses 
+
+* the [*commander*](https://www.npmjs.com/package/commander) package for argument parsing and executing commands, 
```

Comment:
```suggestion
* the [*commander*](https://www.npmjs.com/package/commander) package for argument parsing, 
```

Reply:
I'm pretty sure it also executes the CLI commands. ;)

Reply:
I think (while being true), it is not so important here and implicit with argument parsing :)
But I also see the argument for leaving it there, so just decide what you prefer.